File: Algorithms/constructBinaryTreeFromInorderAndPostorderTraversal.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        return buildTreeDriver(inorder, 0, inorder.length-1, postorder, 0, postorder.length-1);
    }
    TreeNode buildTreeDriver(int[] inorder, int inStart, int inEnd, int[] postorder, int postStart, int postEnd) { 	
    	if (postStart > postEnd || inStart > inEnd) {
    		return null;
    	}
    	if (postStart == postEnd) {
    		return new TreeNode(postorder[postStart]);
    	}
    	int rootVal = postorder[postEnd];
    	int rootIndex = 0;
    	for (int i = inStart; i <= inEnd; ++i)
    	{
    		if (inorder[i] == rootVal) {
    			rootIndex = i;
    			break;
    		}
    	}
    	TreeNode root = new TreeNode(rootVal);
    	root.left = buildTreeDriver(inorder, inStart, rootIndex-1, postorder, postStart, postStart+(rootIndex-inStart)-1);
    	root.right = buildTreeDriver(inorder, rootIndex+1, inEnd, postorder, postStart+(rootIndex-inStart), postEnd-1);
    	return root;
    }
}

File: Algorithms/constructBinaryTreeFromInorderAndPostorderTraversal.py
# Definition for a  binary tree node
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    # @param inorder, a list of integers
    # @param postorder, a list of integers
    # @return a tree node
    def buildTree(self, inorder, postorder):
        return self.buildTreeHelper(inorder, 0, len(inorder)-1, postorder, 0, len(postorder)-1)

    def buildTreeHelper(self, inorder, inStart, inEnd, postorder, postStart, postEnd):
        if inStart > inEnd: return None
        root = TreeNode(postorder[postEnd])
        if inStart == inEnd: return root
        for root_index in range(inStart, inEnd+1):
            if inorder[root_index] == root.val:
                break
        root.left = self.buildTreeHelper(inorder, inStart, root_index-1, 
                postorder, postStart, postStart+root_index-inStart-1)
        root.right = self.buildTreeHelper(inorder, root_index+1, inEnd,
                postorder, postStart+root_index-inStart, postEnd-1)
        return root

